README.md
# Learning Clojure

Learning clojure from https://www.braveclojure.com/clojure-for-the-brave-and-true/ 

## Installation

Install leiningen from https://leiningen.org/. To create a new project use:
```bash
lein new app name_app
```

Install Calva for Visual Studio Code from Extensions, execute REPL.

## Usage

FIXME: explanation

    $ java -jar clojure_learning-0.1.0-standalone.jar [args]

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-03-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-03-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure_learning/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure_learning/compare/0.1.0...0.1.1


src/clojure_learning/core.clj
(ns clojure-learning.core
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

;; Chapter 3: Do Things: A Clojure Crash Course
;; https://www.braveclojure.com/do-things/

;; ----- Exercise 1
;; Use the str, vector, list, hash-map, and hash-set functions.
;; -> str
(str "Hi!, Clojure" " and Calva" " and something")
;; -> vector
(get [8 7 "Hi" {:a 1}] 3)
(vector "Maria" 34.2 "moon")
(conj [3 4 5] 9.8)
;; -> list
(conj '(7 6 2 {:a 1 :b 4}) "Hi")
(nth '(:a :b :c) 1)
(list 1 "two" 3 4 {5 6} {7 8})
;; -> hash-map
(hash-map :one 1 :two 2 :three 3)
(get-in {:one 1 :two {:calva "calva" :clj "clojure"}} [:two :clj])
;; -> functions
(defn function-a
  [name]
  (str "HI " name ", you're so pretty"))
(function-a "Pau")
;; anonymous function
((fn [name] (str "HI, " name)) "Cami")
(#(str "Hi, " %) "Luci")

;; ----- Exercise 2
;; Write a function that takes a number and adds 100 to it.
(defn add100
  [x]
  (+ x 100))

(add100 2)
(add100 200)
(add100 12)

;; ----- Exercise 3
;;  Write a function, dec-maker, that works exactly like the function 
;;  inc-maker except with subtraction:
;;  (def dec9 (dec-maker 9))
;;  (dec9 10)
; => 1
(defn dec-maker
  [n]
  #(- % n))
(def dec9 (dec-maker 9))
(dec9 10)

;; ----- Exercise 4
;; Write a function, mapset, that works like map except the return value is a set:
;; (mapset inc [1 1 2 2])
; => #{2 3}
(defn mapset
  [f numbers]
  (set (map f numbers)))
(mapset inc [1 1 2 2])

;; ----- Exercise 5
;; Create a function that’s similar to symmetrize-body-parts except that it has 
;; to work with weird space aliens with radial symmetry. Instead of two eyes, 
;; arms, legs, and so on, they have five.
(def asym-hobbit-body-parts [{:name "head" :size 3}
                             {:name "left-eye" :size 1}
                             {:name "left-ear" :size 1}
                             {:name "mouth" :size 1}
                             {:name "nose" :size 1}
                             {:name "neck" :size 2}
                             {:name "left-shoulder" :size 3}
                             {:name "left-upper-arm" :size 3}
                             {:name "chest" :size 10}
                             {:name "back" :size 10}
                             {:name "left-forearm" :size 3}
                             {:name "abdomen" :size 6}
                             {:name "left-kidney" :size 1}
                             {:name "left-hand" :size 2}
                             {:name "left-knee" :size 2}
                             {:name "left-thigh" :size 4}
                             {:name "left-lower-leg" :size 3}
                             {:name "left-achilles" :size 1}
                             {:name "left-foot" :size 2}])
(defn radial-matching-part
  [part]
  (map
   (fn [arg]
     {:name (clojure.string/replace (:name part) #"^left-" arg)
      :size (:size part)})
   ["first-" "second-" "third-" "four-" "five-"]))
(defn radial-symmetry-body-parts
  [body-parts]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set (radial-matching-part part))))
          []
          body-parts))
(radial-symmetry-body-parts asym-hobbit-body-parts)

;[{:name "head", :size 3}
; {:name "four-eye", :size 1}
; {:name "five-eye", :size 1}
; {:name "first-eye", :size 1}
; {:name "second-eye", :size 1}
; {:name "third-eye", :size 1}
; {:name "second-ear", :size 1}
; {:name "first-ear", :size 1}
; {:name "five-ear", :size 1}
; {:name "third-ear", :size 1}
; {:name "four-ear", :size 1}
; {:name "mouth", :size 1}
; {:name "nose", :size 1}
; {:name "neck", :size 2}
; {:name "second-shoulder", :size 3}
; {:name "five-shoulder", :size 3}
; {:name "first-shoulder", :size 3}
; {:name "third-shoulder", :size 3}
; {:name "four-shoulder", :size 3}
; {:name "third-upper-arm", :size 3}
; {:name "four-upper-arm", :size 3}
; {:name "second-upper-arm", :size 3}
; {:name "first-upper-arm", :size 3}
; {:name "five-upper-arm", :size 3}
; {:name "chest", :size 10}
; {:name "back", :size 10}
; {:name "first-forearm", :size 3}
; {:name "four-forearm", :size 3}
; {:name "second-forearm", :size 3}
; {:name "five-forearm", :size 3}
; {:name "third-forearm", :size 3}
; {:name "abdomen", :size 6}
; {:name "second-kidney", :size 1}
; {:name "five-kidney", :size 1}
; {:name "third-kidney", :size 1}
; {:name "first-kidney", :size 1}
; {:name "four-kidney", :size 1}
; {:name "first-hand", :size 2}
; {:name "second-hand", :size 2}
; {:name "four-hand", :size 2}
; {:name "third-hand", :size 2}
; {:name "five-hand", :size 2}
; {:name "five-knee", :size 2}
; {:name "four-knee", :size 2}
; {:name "second-knee", :size 2}
; {:name "first-knee", :size 2}
; {:name "third-knee", :size 2}
; {:name "third-thigh", :size 4}
; {:name "five-thigh", :size 4}
; {:name "second-thigh", :size 4}
; {:name "four-thigh", :size 4}
; {:name "first-thigh", :size 4}
; {:name "third-lower-leg", :size 3}
; {:name "second-lower-leg", :size 3}
; {:name "four-lower-leg", :size 3}
; {:name "first-lower-leg", :size 3}
; {:name "five-lower-leg", :size 3}
; {:name "five-achilles", :size 1}
; {:name "second-achilles", :size 1}
; {:name "four-achilles", :size 1}
; {:name "third-achilles", :size 1}
; {:name "first-achilles", :size 1}
; {:name "third-foot", :size 2}
; {:name "five-foot", :size 2}
; {:name "four-foot", :size 2}
; {:name "first-foot", :size 2}
; {:name "second-foot", :size 2}]

;; ----- Exercise 6
;; Create a function that generalizes symmetrize-body-parts and the function 
;; you created in Exercise 5. The new function should take a collection of body 
;; parts and the number of matching body parts to add. If you’re completely 
;; new to Lisp languages and functional programming, it probably won’t be obvious 
;; how to do this. If you get stuck, just move on to the next chapter and 
;; revisit the problem later.

(defn generalize-matching-part
  [part n]
  (map
   (fn [arg]
     {:name (clojure.string/replace (:name part) #"^left-" (str arg "-"))
      :size (:size part)})
   (vec (range 1 (inc n)))))

(defn generalize-symmetry-body-parts
  [body-parts n]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set (generalize-matching-part part n))))
          []
          body-parts))
(generalize-symmetry-body-parts asym-hobbit-body-parts 5)



src/clojure_learning/chapter4.clj
;; Chapter 4: Core Functions in Depth

(def food-journal
  [{:month 1 :day 1 :human 5.3 :critter 2.3}
   {:month 1 :day 2 :human 5.1 :critter 2.0}
   {:month 2 :day 1 :human 4.9 :critter 2.1}
   {:month 2 :day 2 :human 5.0 :critter 2.5}
   {:month 3 :day 1 :human 4.2 :critter 3.3}
   {:month 3 :day 2 :human 4.0 :critter 3.8}
   {:month 4 :day 1 :human 3.7 :critter 3.9}
   {:month 4 :day 2 :human 3.7 :critter 3.6}])

;; take while -. take n elements from a collection
(take-while #(< (:month %) 3) food-journal)
;; => ({:month 1, :day 1, :human 5.3, :critter 2.3}
;;     {:month 1, :day 2, :human 5.1, :critter 2.0}
;;     {:month 2, :day 1, :human 4.9, :critter 2.1}
;;     {:month 2, :day 2, :human 5.0, :critter 2.5})

;; drop while -. drop n elements from a collection
(drop-while #(< (:month %) 3) food-journal)
;; => ({:month 3, :day 1, :human 4.2, :critter 3.3}
;;     {:month 3, :day 2, :human 4.0, :critter 3.8}
;;     {:month 4, :day 1, :human 3.7, :critter 3.9}
;;     {:month 4, :day 2, :human 3.7, :critter 3.6})

;; Those are more efficient than: 
;; filter, filter elements from a collection
(filter #(< (:month %) 3) food-journal)
;; => ({:month 1, :day 1, :human 5.3, :critter 2.3}
;;     {:month 1, :day 2, :human 5.1, :critter 2.0}
;;     {:month 2, :day 1, :human 4.9, :critter 2.1}
;;     {:month 2, :day 2, :human 5.0, :critter 2.5})

;; some, if collection contains any value that etst true
(some #(> (:critter %) 5) food-journal)
;; => nil

(some #(> (:critter %) 3) food-journal) ; <- only return true or false
;; => true

(some #(and (> (:critter %) 5) %) food-journal) ; <- return entry
;; => {:month 3, :day 1, :human 4.2, :critter 3.3}

;; sort to sort elements, and sort-by to apply function to the elements of a sequence
(sort [3 5 2 1 4.6])
;; => (1 2 3 4.6 5)

(sort-by count ["aaa" "c" "bb"])
;; => ("c" "bb" "aaa")

;; concat to appends members to one sequence to the end of another
(concat ["hi" 3.2 4] [4 1])
;; => ("hi" 3.2 4 4 1)

(def vampire-database
  {0 {:makes-blood-puns? false, :has-pulse? true  :name "McFishwich"}
   1 {:makes-blood-puns? false, :has-pulse? true  :name "McMackson"}
   2 {:makes-blood-puns? true,  :has-pulse? false :name "Damon Salvatore"}
   3 {:makes-blood-puns? true,  :has-pulse? true  :name "Mickey Mouse"}})

(defn vampire-related-details
  [social-security-number]
  (Thread/sleep 1000)
  (get vampire-database social-security-number))

(defn vampire?
  [record]
  (and (:makes-blood-puns? record)
       (not (:has-pulse? record))
       record))

(defn identify-vampire
  [social-security-numbers]
  (first (filter vampire?
                 (map vampire-related-details social-security-numbers))))

(time (vampire-related-details 0)) ; time how much time it takes to run, only REPL
; "Elapsed time: 1000.2303 msecs"
;; => {:makes-blood-puns? false, :has-pulse? true, :name "McFishwich"}
;; 
(time (identify-vampire (range 0 1000000)))
;; => {:makes-blood-puns? true, :has-pulse? false, :name "Damon Salvatore"}


(time (def mapped-details (map vampire-related-details (range 0 1000000))))
; "Elapsed time: 0.058597 msecs"
;; => #'user/mapped-details

;; Infinite sequenes
(concat (take 9 (repeat "la")) ["Con altura!"]) ; using repeat
;; => ("la" "la" "la" "la" "la" "la" "la" "la" "la" "Con altura!")

(take 3 (repeatedly (fn [] (rand-int 10)))) ; using repeatedly
;; => (7 5 2)

(defn even-numbers
  ([] (even-numbers 0))
  ([n] (cons n (lazy-seq (even-numbers (+ n 2))))))
(take 10 (even-numbers))
;; => (0 2 4 6 8 10 12 14 16 18)

(cons 0 '(2 5 8)) ; append element to list
;; => (0 2 5 8)

(empty? []) ; about the whole, like count and every?
;; => true

(empty? ["no!"])
;; => false

;; into - convert return value seq back into original value
(map identity {:sunlight-reaction "Glitter!"})
;; => ([:sunlight-reaction "Glitter!"])

(into {} (map identity {:sunlight-reaction "Glitter!"}))
;; => {:sunlight-reaction "Glitter!"}

(into {:favorite-emotion "gloomy"} [[:sunlight-reaction "Glitter!"]])
;; => {:favorite-emotion "gloomy", :sunlight-reaction "Glitter!"}

(into ["cherry"] '("pine" "spruce"))
;; => ["cherry" "pine" "spruce"]

;; conj add elements but is different from into
;; conj takes a rest parameter
;; into takes a seq data structure
(conj [0] [1])
;; => [0 [1]]

(into [0] [1])
;; => [0 1]

(conj [0] 1)
;; => [0 1]

(conj [0] 1 2 3 4 5)
;; => [0 1 2 3 4 5]

;; define conj in terms of into
(defn my-conj
  [target & additions]
  (into target additions))
(my-conj [0] 1 2 3) ; -> define conj in base of into
;; => [0 1 2 3]

;; define into in terms of conj
(defn my-into
  [target additions]
  (apply conj target additions)) ; apply to every element of additions
(my-into [0] [1 2 3])
;; => [0 1 2 3]


;; apply, when use max to get the maximun of a seq, it only works in single elements
;; can use apply to a data structure to get the max
(max 0 1 2) ; it works for numbers
;; => 2
 
;(max [0 1 2]) ; return entire data structure
;; => [0 1 2]

(apply max [0 1 2]) ; use apply to get max in data structures
;; => 2

;; partial
(def add10 (partial + 10))
(add10 3)
;; => 13

(def add-missing-elements
  (partial conj ["water" "earth" "air"]))
(add-missing-elements "unobtainium" "adamantium")
;; => ["water" "earth" "air" "unobtainium" "adamantium"]

(defn my-partial
  [partialized-fn & args]
  (fn [& more-args]
    (apply partialized-fn (into args more-args))))
(def add20 (my-partial + 20))
(add20 3)
;; => 23

;; WHEN to use partial? when find repeating the same combination of function
;; and arguments in many different contexts
(defn lousy-logger
  [log-level message]
  (condp = log-level
    :warn (clojure.string/lower-case message)
    :emergency (clojure.string/upper-case message)))
(def warn (partial lousy-logger :warn))
(warn "Red light ahead")
;; => "red light ahead"

(def emergency (partial lousy-logger :emergency))
(emergency "Red light ahead")
;; => "RED LIGHT AHEAD"

;; complement
(defn identify-humans
  [social-security-numbers]
  (filter #(not (vampire? %))
          (map vampire-related-details social-security-numbers)))

(def not-vampire? (complement vampire?)) ; complement
(defn identify-humans
  [social-security-numbers]
  (filter not-vampire?
          (map vampire-related-details social-security-numbers)))















src/clojure_learning/chapter3.clj
;; Chapter 3: Do Things: A Clojure Crash Course
;; https://www.braveclojure.com/do-things/

;; ----- Exercise 1
;; Use the str, vector, list, hash-map, and hash-set functions.
;; -> str
(str "Hi!, Clojure" " and Calva" " and something")
;; => "Hi!, Clojure and Calva and something"

;; -> vector
(get [8 7 "Hi" {:a 1}] 3)
;; => {:a 1}

(vector "Maria" 34.2 "moon")
;; => ["Maria" 34.2 "moon"]

(conj [3 4 5] 9.8)
;; => [3 4 5 9.8]

;; -> list
(conj '(7 6 2 {:a 1 :b 4}) "Hi")
;; => ("Hi" 7 6 2 {:a 1, :b 4})

(nth '(:a :b :c) 1)
;; => :b

(list 1 "two" 3 4 {5 6} {7 8})
;; => (1 "two" 3 4 {5 6} {7 8})

;; -> hash-map
(hash-map :one 1 :two 2 :three 3)
;; => {:one 1, :three 3, :two 2}

(get-in {:one 1 :two {:calva "calva" :clj "clojure"}} [:two :clj])
;; => "clojure"

;; -> functions
(defn function-a
  [name]
  (str "HI " name ", you're so pretty"))
(function-a "Pau")
;; => "HI Pau, you're so pretty"

;; anonymous function
((fn [name] (str "HI, " name)) "Cami")
;; => "HI, Cami"

(#(str "Hi, " %) "Luci")
;; => "Hi, Luci"


;; ----- Exercise 2
;; Write a function that takes a number and adds 100 to it.
(defn add100
  [x]
  (+ x 100))

(add100 2)
;; => 102

(add100 200)
;; => 300

(add100 12)
;; => 112


;; ----- Exercise 3
;;  Write a function, dec-maker, that works exactly like the function 
;;  inc-maker except with subtraction:
;;  (def dec9 (dec-maker 9))
;;  (dec9 10)
; => 1
(defn dec-maker
  [n]
  #(- % n))
(def dec9 (dec-maker 9))
(dec9 10)
;; => 1


;; ----- Exercise 4
;; Write a function, mapset, that works like map except the return value is a set:
;; (mapset inc [1 1 2 2])
; => #{2 3}
(defn mapset
  [f numbers]
  (set (map f numbers)))
(mapset inc [1 1 2 2])
;; => #{3 2}


;; ----- Exercise 5
;; Create a function that’s similar to symmetrize-body-parts except that it has 
;; to work with weird space aliens with radial symmetry. Instead of two eyes, 
;; arms, legs, and so on, they have five.
(def asym-hobbit-body-parts [{:name "head" :size 3}
                             {:name "left-eye" :size 1}
                             {:name "left-ear" :size 1}
                             {:name "mouth" :size 1}
                             {:name "nose" :size 1}
                             {:name "neck" :size 2}
                             {:name "left-shoulder" :size 3}
                             {:name "left-upper-arm" :size 3}
                             {:name "chest" :size 10}
                             {:name "back" :size 10}
                             {:name "left-forearm" :size 3}
                             {:name "abdomen" :size 6}
                             {:name "left-kidney" :size 1}
                             {:name "left-hand" :size 2}
                             {:name "left-knee" :size 2}
                             {:name "left-thigh" :size 4}
                             {:name "left-lower-leg" :size 3}
                             {:name "left-achilles" :size 1}
                             {:name "left-foot" :size 2}])
(defn radial-matching-part
  [part]
  (map
   (fn [arg]
     {:name (clojure.string/replace (:name part) #"^left-" arg)
      :size (:size part)})
   ["first-" "second-" "third-" "four-" "five-"]))
(defn radial-symmetry-body-parts
  [body-parts]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set (radial-matching-part part))))
          []
          body-parts))
(radial-symmetry-body-parts asym-hobbit-body-parts)
;; => [{:name "head", :size 3}
;;     {:name "four-eye", :size 1}
;;     {:name "five-eye", :size 1}
;;     {:name "first-eye", :size 1}
;;     {:name "second-eye", :size 1}
;;     {:name "third-eye", :size 1}
;;     {:name "second-ear", :size 1}
;;     {:name "first-ear", :size 1}
;;     {:name "five-ear", :size 1}
;;     {:name "third-ear", :size 1}
;;     {:name "four-ear", :size 1}
;;     {:name "mouth", :size 1}
;;     {:name "nose", :size 1}
;;     {:name "neck", :size 2}
;;     {:name "second-shoulder", :size 3}
;;     {:name "five-shoulder", :size 3}
;;     {:name "first-shoulder", :size 3}
;;     {:name "third-shoulder", :size 3}
;;     {:name "four-shoulder", :size 3}
;;     {:name "third-upper-arm", :size 3}
;;     {:name "four-upper-arm", :size 3}
;;     {:name "second-upper-arm", :size 3}
;;     {:name "first-upper-arm", :size 3}
;;     {:name "five-upper-arm", :size 3}
;;     {:name "chest", :size 10}
;;     {:name "back", :size 10}
;;     {:name "first-forearm", :size 3}
;;     {:name "four-forearm", :size 3}
;;     {:name "second-forearm", :size 3}
;;     {:name "five-forearm", :size 3}
;;     {:name "third-forearm", :size 3}
;;     {:name "abdomen", :size 6}
;;     {:name "second-kidney", :size 1}
;;     {:name "five-kidney", :size 1}
;;     {:name "third-kidney", :size 1}
;;     {:name "first-kidney", :size 1}
;;     {:name "four-kidney", :size 1}
;;     {:name "first-hand", :size 2}
;;     {:name "second-hand", :size 2}
;;     {:name "four-hand", :size 2}
;;     {:name "third-hand", :size 2}
;;     {:name "five-hand", :size 2}
;;     {:name "five-knee", :size 2}
;;     {:name "four-knee", :size 2}
;;     {:name "second-knee", :size 2}
;;     {:name "first-knee", :size 2}
;;     {:name "third-knee", :size 2}
;;     {:name "third-thigh", :size 4}
;;     {:name "five-thigh", :size 4}
;;     {:name "second-thigh", :size 4}
;;     {:name "four-thigh", :size 4}
;;     {:name "first-thigh", :size 4}
;;     {:name "third-lower-leg", :size 3}
;;     {:name "second-lower-leg", :size 3}
;;     {:name "four-lower-leg", :size 3}
;;     {:name "first-lower-leg", :size 3}
;;     {:name "five-lower-leg", :size 3}
;;     {:name "five-achilles", :size 1}
;;     {:name "second-achilles", :size 1}
;;     {:name "four-achilles", :size 1}
;;     {:name "third-achilles", :size 1}
;;     {:name "first-achilles", :size 1}
;;     {:name "third-foot", :size 2}
;;     {:name "five-foot", :size 2}
;;     {:name "four-foot", :size 2}
;;     {:name "first-foot", :size 2}
;;     {:name "second-foot", :size 2}]

;; ----- Exercise 6
;; Create a function that generalizes symmetrize-body-parts and the function 
;; you created in Exercise 5. The new function should take a collection of body 
;; parts and the number of matching body parts to add. If you’re completely 
;; new to Lisp languages and functional programming, it probably won’t be obvious 
;; how to do this. If you get stuck, just move on to the next chapter and 
;; revisit the problem later.

(defn generalize-matching-part
  [part n]
  (map
   (fn [arg]
     {:name (clojure.string/replace (:name part) #"^left-" (str arg "-"))
      :size (:size part)})
   (vec (range 1 (inc n)))))

(defn generalize-symmetry-body-parts
  [body-parts n]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set (generalize-matching-part part n))))
          []
          body-parts))
(generalize-symmetry-body-parts asym-hobbit-body-parts 5)

;; => [{:name "head", :size 3}
;;     {:name "5-eye", :size 1}
;;     {:name "3-eye", :size 1}
;;     {:name "2-eye", :size 1}
;;     {:name "1-eye", :size 1}
;;     {:name "4-eye", :size 1}
;;     {:name "3-ear", :size 1}
;;     {:name "1-ear", :size 1}
;;     {:name "2-ear", :size 1}
;;     {:name "4-ear", :size 1}
;;     {:name "5-ear", :size 1}
;;     {:name "mouth", :size 1}
;;     {:name "nose", :size 1}
;;     {:name "neck", :size 2}
;;     {:name "5-shoulder", :size 3}
;;     {:name "4-shoulder", :size 3}
;;     {:name "2-shoulder", :size 3}
;;     {:name "1-shoulder", :size 3}
;;     {:name "3-shoulder", :size 3}
;;     {:name "4-upper-arm", :size 3}
;;     {:name "5-upper-arm", :size 3}
;;     {:name "3-upper-arm", :size 3}
;;     {:name "2-upper-arm", :size 3}
;;     {:name "1-upper-arm", :size 3}
;;     {:name "chest", :size 10}
;;     {:name "back", :size 10}
;;     {:name "2-forearm", :size 3}
;;     {:name "4-forearm", :size 3}
;;     {:name "1-forearm", :size 3}
;;     {:name "5-forearm", :size 3}
;;     {:name "3-forearm", :size 3}
;;     {:name "abdomen", :size 6}
;;     {:name "2-kidney", :size 1}
;;     {:name "4-kidney", :size 1}
;;     {:name "5-kidney", :size 1}
;;     {:name "3-kidney", :size 1}
;;     {:name "1-kidney", :size 1}
;;     {:name "5-hand", :size 2}
;;     {:name "2-hand", :size 2}
;;     {:name "4-hand", :size 2}
;;     {:name "1-hand", :size 2}
;;     {:name "3-hand", :size 2}
;;     {:name "3-knee", :size 2}
;;     {:name "1-knee", :size 2}
;;     {:name "5-knee", :size 2}
;;     {:name "2-knee", :size 2}
;;     {:name "4-knee", :size 2}
;;     {:name "3-thigh", :size 4}
;;     {:name "2-thigh", :size 4}
;;     {:name "5-thigh", :size 4}
;;     {:name "4-thigh", :size 4}
;;     {:name "1-thigh", :size 4}
;;     {:name "1-lower-leg", :size 3}
;;     {:name "4-lower-leg", :size 3}
;;     {:name "5-lower-leg", :size 3}
;;     {:name "3-lower-leg", :size 3}
;;     {:name "2-lower-leg", :size 3}
;;     {:name "2-achilles", :size 1}
;;     {:name "5-achilles", :size 1}
;;     {:name "3-achilles", :size 1}
;;     {:name "4-achilles", :size 1}
;;     {:name "1-achilles", :size 1}
;;     {:name "2-foot", :size 2}
;;     {:name "5-foot", :size 2}
;;     {:name "3-foot", :size 2}
;;     {:name "1-foot", :size 2}
;;     {:name "4-foot", :size 2}]

